pkg/controller/controller.go

```diff
@@ -0,0 +1,272 @@
+	c.processingMutex.Lock()
+	defer c.processingMutex.Unlock()
+	c.removeServiceWatcher(clusterID)
+}
+
+func (c *LightHouseController) startNewServiceWatcher(clusterID string, clientSet kubernetes.Interface) {
+	serviceInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options meta_v1.ListOptions) (runtime.Object, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options meta_v1.ListOptions) (watch.Interface, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Service{},
+		0,
+		cache.Indexers{},
+	)
+	remoteCluster := &RemoteCluster{
+		stopCh:          make(chan struct{}),
+		ClusterID:       clusterID,
+		serviceInformer: serviceInformer,
+		queue:           workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+	c.processingMutex.Lock()
+	c.remoteClusters[clusterID] = remoteCluster
+	c.processingMutex.Unlock()
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go c.Run(remoteCluster)
+}
+
+func (c *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := c.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(c.remoteClusters, clusterID)
+	}
+}
+
+func (rc *RemoteCluster) close() {
+	close(rc.stopCh)
+	rc.queue.ShutDown()
+	klog.Infof(" watcher queue shutdown for cluster: %s", rc.ClusterID)
+}
+
+func (c *LightHouseController) Run(remoteCluster *RemoteCluster) {
+	defer utilruntime.HandleCrash()
+
+	go remoteCluster.serviceInformer.Run(remoteCluster.stopCh)
+
+	if !cache.WaitForCacheSync(remoteCluster.stopCh, remoteCluster.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go c.runWorker(remoteCluster)
+}
+
+// HasSynced is required for the cache.Controller interface.
+func (c *RemoteCluster) HasSynced() bool {
+	return c.serviceInformer.HasSynced()
+}
+
+func (c *LightHouseController) runWorker(remoteCluster *RemoteCluster) {
+	for {
+		key, shutdown := remoteCluster.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for cluster \"%s\" stopped", remoteCluster.ClusterID)
+			return
+		}
+		defer remoteCluster.queue.Done(key)
+
+		obj, exists, err := remoteCluster.serviceInformer.GetIndexer().GetByKey(key.(string))
+
+		if err != nil {
+			klog.Errorf("Error processing %s (will retry): %v", key, err)
+			// requeue the item to work on later
+			remoteCluster.queue.AddRateLimited(key)
+		}
+		if !exists {
+			c.ObjectDeleted(obj, remoteCluster.ClusterID)
```

Comment:
`obj` will be nil here so `ObjectDeleted` will cause a panic. It looks like you only need the service `Name` which you can extract from the `key`.